### PR TITLE
Change PDP listener interface to multiple methods rather than structured payload

### LINKED ISSUE #55: PDP Listener should be an interface with separate methods
The PDP Listener instead of taking an "OperationType" parameter should just have a bunch of different methods "receiveAddRoots", "receiveProvePossession" etc.

We should make an FRC to canonicalize this interface type.

Note the interface is going to be very similar to the DDO notify methods just a little more rich since fault checking isn't coming from cron since we're doing PDP correctly the first time.

### PR DESCRIPTION
**Pull request by anorth**
Closes #55.

### CODE REVIEW

**Comment by anorth** [comment]
The thrust of the main change is done, but tests not updated yet.

**Review by ZenGround0 — COMMENTED** [review]
This looks great

File: contracts/src/PDPRecordKeeper.sol [hunk 1]
```diff
@@ -35,13 +44,29 @@ contract PDPRecordKeeper is PDPListener {
...
+    function proofSetCreated(uint256 proofSetId, address creator) external onlyPDPService {
+        receiveProofSetEvent(proofSetId, OperationType.CREATE, abi.encode(creator));
+    }
+
+    function rootsAdded(uint256 proofSetId, uint256 firstAdded, PDPService.RootData[] memory rootData) external onlyPDPService {
+        receiveProofSetEvent(proofSetId, OperationType.ADD, abi.encode(firstAdded, rootData));
+    }
+
+    function rootsScheduledRemove(uint256 proofSetId, uint256[] memory rootIds) external onlyPDPService {
+        receiveProofSetEvent(proofSetId, OperationType.REMOVE, abi.encode(rootIds));
```
Comment (anorth) [inline contracts/src/PDPRecordKeeper.sol:57]: Note the total delta was dropped from the event payload here, following our discussion about it re #46

File: contracts/src/PDPRecordKeeper.sol [hunk 2]
```diff
@@ -35,13 +44,29 @@ contract PDPRecordKeeper is PDPListener {
...
+    }
+
+    function posessionProven(uint256 proofSetId, uint256 challengedLeafCount, uint256 seed, uint256 challengeCount) external onlyPDPService {
```
Comment (anorth) [inline contracts/src/PDPRecordKeeper.sol:60]: Added challenge count to the payload here.

**Review by ZenGround0 — APPROVED** [review]

File: contracts/src/PDPService.sol [hunk 3]
```diff
@@ -178,14 +168,14 @@ contract PDPService {
         proofSetOwner[setId] = msg.sender;
         proofSetListener[setId] = listenerAddr;
 
-        bytes memory extraData = abi.encode(msg.sender);
-        _notifyListener(setId, listenerAddr, PDPListener.OperationType.CREATE, extraData);
+        if (listenerAddr != address(0)) {
```
[the remark this answers is not in the document]
Comment (anorth) [reply contracts/src/PDPService.sol]: As discussed, this supports a proof set with no listener (e.g. for off-chain verification).

File: contracts/src/PDPRecordKeeper.sol [hunk 4]
```diff
@@ -11,18 +11,27 @@ contract PDPRecordKeeper is PDPListener {
...
+    enum OperationType {
+        NONE,
+        CREATE,
+        ADD,
+        REMOVE,
+        PROVE_POSSESSION,
+        DELETE
+    }
+
     // Struct to store event details
```
[the remark this answers is not in the document]
Comment (anorth) [reply contracts/src/PDPRecordKeeper.sol]: I've added the proof set ID field. I think the epoch should remain: this simple service accumulates events in state so they can be queried later.

**Comment by anorth** [comment]
@ZenGround0 I did a big rebase so you should look through this again, but I think it's good to go now. 

There's now a nextProvingPeriod event method, and I added a leafCount parameter to it.

**Review by ZenGround0 — APPROVED** [review]

**Pull request merged**